misc/extension_api_validation/4.2-stable.expected

```diff
@@ -357,3 +356,14 @@ Validate extension JSON: Error: Field 'classes/RenderingServer/methods/canvas_it
 Validate extension JSON: Error: Field 'classes/RenderingServer/methods/canvas_item_add_rect/arguments': size changed value in new API, from 3 to 4.
 
 Optional arguments added. Compatibility methods registered.
+
+
+GH-90993
+--------
+Validate extension JSON: Error: Field 'classes/RenderingDevice/methods/draw_list_begin/arguments/9': type changed value in new API, from "Array" to "int".
+Validate extension JSON: Error: Field 'classes/RenderingDevice/methods/draw_list_begin/arguments/9': default_value changed value in new API, from "Array[RID]([])" to "0".
+Validate extension JSON: Error: Field 'classes/RenderingDevice/methods/draw_list_begin/arguments/9': type changed value in new API, from "typedarray::RID" to "int".
```

Comment:
Only one of these should be kept I think, the type part

Reply:
The last two are complaints from 4.2 -> 4.3
But the last one is a complaint from 4.1 (or 4.0?) -> 4.3

I was a bit confused by it because the same issue should've happened when going 4.1 -> 4.2

Reply:
Hm there seems to be a clash with #84976, which is a 4.2->4.3 change that _removed_ the 9th argument to `draw_list_begin`.

Now this PR adds a new 9th argument with a different type.

I think it's the first time we have this case, not sure if the script we use to check for compatibility issues is able to handle that. CC @dsnopek

Reply:
Since both changes were in 4.3, I would expect it to treat this like the 9th argument changed type, completely ignoring the previous removal of the argument.

The weird thing is these two lines:

```
Validate extension JSON: Error: Field 'classes/RenderingDevice/methods/draw_list_begin/arguments/9': type changed value in new API, from "Array" to "int".
Validate extension JSON: Error: Field 'classes/RenderingDevice/methods/draw_list_begin/arguments/9': type changed value in new API, from "typedarray::RID" to "int".
```

I'm not sure how it can change from both `Array` and `typedarray::RID`.

Looking at the [reference `extension_api.json`](https://raw.githubusercontent.com/godotengine/godot-cpp/godot-4.2-stable/gdextension/extension_api.json), it seems like `typedarray::RID` is the correct one.

What happens if you remove this line from the `4.2-stable.expected`:

```
Validate extension JSON: Error: Field 'classes/RenderingDevice/methods/draw_list_begin/arguments/9': type changed value in new API, from "Array" to "int".
```

Reply:
If I remove the first entry and only leave:

```
GH-90993
--------
Validate extension JSON: Error: Field 'classes/RenderingDevice/methods/draw_list_begin/arguments/9': default_value changed value in new API, from "Array[RID]([])" to "0".
Validate extension JSON: Error: Field 'classes/RenderingDevice/methods/draw_list_begin/arguments/9': type changed value in new API, from "typedarray::RID" to "int".

draw_list_begin added a new optional debug argument called breadcrumb.
There used to be an RID argument as arg #9 which was removed in GH-84976 but now we're adding a new one in the same location.
No compatibility method was registered because the one created for GH-84976 is already handling this case.
```

Then the CI still fails. The output is:

```
Godot Engine v4.3.beta.custom_build.2446b419d (2024-05-18 16:21:37 UTC) - https://godotengine.org

2024-05-18 17:07:59 URL:https://raw.githubusercontent.com/godotengine/godot-cpp/godot-4.1-stable/gdextension/extension_api.json [5318978/5318978] -> "/tmp/tmp.exex1R1Xhb" [1]

ERROR: New API lacks base array: enums
   at: compare_dict_array (core/extension/extension_api_dump.cpp:1362)
ERROR: Validate extension JSON: Missing field in current API 'classes/RenderingDevice/methods/draw_list_end': arguments. This is a bug.
   at: compare_dict_array (core/extension/extension_api_dump.cpp:1443)
ERROR: Validate extension JSON: Missing field in current API 'classes/RenderingDevice/methods/compute_list_begin': arguments. This is a bug.
   at: compare_dict_array (core/extension/extension_api_dump.cpp:1443)
ERROR: Validate extension JSON: Missing field in current API 'classes/RenderingDevice/methods/compute_list_end': arguments. This is a bug.
   at: compare_dict_array (core/extension/extension_api_dump.cpp:1443)
Compatibility to 4.1-stable is broken in the following ways:
Validate extension JSON: API was removed: classes/OpenXRAction
Validate extension JSON: API was removed: classes/OpenXRActionMap
Validate extension JSON: API was removed: classes/OpenXRActionSet
Validate extension JSON: API was removed: classes/OpenXRHand
Validate extension JSON: API was removed: classes/OpenXRInteractionProfile
Validate extension JSON: API was removed: classes/OpenXRInterface
Validate extension JSON: API was removed: classes/OpenXRIPBinding
Validate extension JSON: API was removed: classes/WebXRInterface
Godot Engine v4.3.beta.custom_build.2446b419d (2024-05-18 16:21:37 UTC) - https://godotengine.org

2024-05-18 17:12:27 URL:https://raw.githubusercontent.com/godotengine/godot-cpp/godot-4.0-stable/gdextension/extension_api.json [5175722/5175722] -> "/tmp/tmp.qkBYdofWdn" [1]


ERROR: New API lacks base array: enums
   at: compare_dict_array (core/extension/extension_api_dump.cpp:1362)
ERROR: Validate extension JSON: Missing field in current API 'classes/RenderingDevice/methods/draw_list_end': arguments. This is a bug.
   at: compare_dict_array (core/extension/extension_api_dump.cpp:1443)
ERROR: Validate extension JSON: Missing field in current API 'classes/RenderingDevice/methods/compute_list_begin': arguments. This is a bug.
   at: compare_dict_array (core/extension/extension_api_dump.cpp:1443)
ERROR: Validate extension JSON: Missing field in current API 'classes/RenderingDevice/methods/compute_list_end': arguments. This is a bug.
   at: compare_dict_array (core/extension/extension_api_dump.cpp:1443)
Compatibility to 4.0-stable is broken in the following ways:
Validate extension JSON: API was removed: classes/OpenXRAction
Validate extension JSON: API was removed: classes/OpenXRActionMap
Validate extension JSON: API was removed: classes/OpenXRActionSet
Validate extension JSON: API was removed: classes/OpenXRHand
Validate extension JSON: API was removed: classes/OpenXRInteractionProfile
Validate extension JSON: API was removed: classes/OpenXRInterface
Validate extension JSON: API was removed: classes/OpenXRIPBinding
Validate extension JSON: API was removed: classes/WebXRInterface
Validate extension JSON: Error: Field 'classes/RenderingDevice/methods/draw_list_begin/arguments/9': type changed value in new API, from "Array" to "int".
Godot Engine v4.3.beta.custom_build.2446b419d (2024-05-18 16:21:37 UTC) - https://godotengine.org

2024-05-18 17:12:29 URL:https://raw.githubusercontent.com/godotengine/godot-cpp/godot-4.2-stable/gdextension/extension_api.json [5501700/5501700] -> "/tmp/tmp.vlgKzMe49U" [1]

ERROR: Validate extension JSON: Missing field in current API 'classes/RenderingDevice/methods/draw_list_end': arguments. This is a bug.
   at: compare_dict_array (core/extension/extension_api_dump.cpp:1443)
ERROR: Validate extension JSON: Missing field in current API 'classes/RenderingDevice/methods/compute_list_begin': arguments. This is a bug.
   at: compare_dict_array (core/extension/extension_api_dump.cpp:1443)
ERROR: Validate extension JSON: Missing field in current API 'classes/RenderingDevice/methods/compute_list_end': arguments. This is a bug.
   at: compare_dict_array (core/extension/extension_api_dump.cpp:1443)
The following validation errors no longer occur (compared to 4.2-stable):
Validate extension JSON: Error: Field 'classes/WebXRInterface/methods/get_input_source_tracker/return_value': type changed value in new API, from "XRPositionalTracker" to "XRControllerTracker".
Compatibility to 4.2-stable is broken in the following ways:
Validate extension JSON: API was removed: classes/OpenXRAction
Validate extension JSON: API was removed: classes/OpenXRActionMap
Validate extension JSON: API was removed: classes/OpenXRActionSet
Validate extension JSON: API was removed: classes/OpenXRAPIExtension
Validate extension JSON: API was removed: classes/OpenXRExtensionWrapperExtension
Validate extension JSON: API was removed: classes/OpenXRHand
Validate extension JSON: API was removed: classes/OpenXRInteractionProfile
Validate extension JSON: API was removed: classes/OpenXRInteractionProfileMetadata
Validate extension JSON: API was removed: classes/OpenXRInterface
Validate extension JSON: API was removed: classes/OpenXRIPBinding
Validate extension JSON: API was removed: classes/WebXRInterface
```

**[AS I SAID](https://github.com/godotengine/godot/pull/90993#discussion_r1605091726)** this error appears in validation from [4.0](https://raw.githubusercontent.com/godotengine/godot-cpp/godot-4.0-stable/gdextension/extension_api.json) -> 4.3

If I restore the
`Validate extension JSON: Error: Field 'classes/RenderingDevice/methods/draw_list_begin/arguments/9': type changed value in new API, from "Array" to "int".` then 4.0 -> 4.3 passes correctly, but [4.2](https://raw.githubusercontent.com/godotengine/godot-cpp/godot-4.2-stable/gdextension/extension_api.json) -> 4.3 will say:

```
The following validation errors no longer occur (compared to 4.2-stable):
Validate extension JSON: Error: Field 'classes/RenderingDevice/methods/draw_list_begin/arguments/9': type changed value in new API, from "Array" to "int".
```

Reply:
In [godot-4.0-stable/gdextension/extension_api.json](https://raw.githubusercontent.com/godotengine/godot-cpp/godot-4.0-stable/gdextension/extension_api.json):

```json
{
	"name": "storage_textures",
	"type": "Array",
	"default_value": "Array[RID]([])"
}
```

In [godot-4.1-stable/gdextension/extension_api.json](https://raw.githubusercontent.com/godotengine/godot-cpp/godot-4.1-stable/gdextension/extension_api.json) & [godot-4.2-stable/gdextension/extension_api.json](https://raw.githubusercontent.com/godotengine/godot-cpp/godot-4.2-stable/gdextension/extension_api.json):

```json
{
	"name": "storage_textures",
	"type": "typedarray::RID",
	"default_value": "Array[RID]([])"
}
```

Reply:
I'll look into it myself, I think Matias' time is better spent in rendering issues than in debugging our CI API compat testing framework.

Reply:
IMO our validation scripts are buggy and need to be fixed to properly handle APIs changing multiple times across versions.

But for now this should be an ok workaround to silence both the error if not adding the line, and the false positive warning when adding it in the `4.2` file.

```diff
diff --git a/misc/extension_api_validation/4.0-stable_4.1-stable.expected b/misc/extension_api_validation/4.0-stable_4.1-stable.expected
index 5c3bf07fb2..7388c90e2e 100644
--- a/misc/extension_api_validation/4.0-stable_4.1-stable.expected
+++ b/misc/extension_api_validation/4.0-stable_4.1-stable.expected
@@ -349,3 +349,11 @@ Validate extension JSON: Error: Hash changed for 'classes/EditorUndoRedoManager/
 Validate extension JSON: Error: Hash changed for 'classes/UndoRedo/methods/create_action', from 0AEC1BFC to E87757EB. This means that the function has changed and no compatibility function was provided.
 
 Added a optional parameters with default values. No adjustments should be necessary.
+
+
+GH-90993
+--------
+Validate extension JSON: Error: Field 'classes/RenderingDevice/methods/draw_list_begin/arguments/9': type changed value in new API, from "Array" to "int".
+
+The change to `int` actually happened from 4.2 to 4.3, but our current validation script doesn't handle well the case where an argument changed type twice in different versions.
+This argument already changed in 4.1 in GH-76418, so the script gets confused.
diff --git a/misc/extension_api_validation/4.2-stable.expected b/misc/extension_api_validation/4.2-stable.expected
index 54f95cde58..cf80419cf9 100644
--- a/misc/extension_api_validation/4.2-stable.expected
+++ b/misc/extension_api_validation/4.2-stable.expected
@@ -367,7 +367,6 @@ Optional arguments added. Compatibility methods registered.
 
 GH-90993
 --------
-Validate extension JSON: Error: Field 'classes/RenderingDevice/methods/draw_list_begin/arguments/9': type changed value in new API, from "Array" to "int".
 Validate extension JSON: Error: Field 'classes/RenderingDevice/methods/draw_list_begin/arguments/9': default_value changed value in new API, from "Array[RID]([])" to "0".
 Validate extension JSON: Error: Field 'classes/RenderingDevice/methods/draw_list_begin/arguments/9': type changed value in new API, from "typedarray::RID" to "int".
 
```

Reply:
Done!!!

I've applied your patch, rebased to latest master, and pushed.

Now all CI passes!